### PR TITLE
Implement up --continue-on-error 

### PR DESCRIPTION
**Pull request by tgummerer**
Similar to destroy --continue-on-error, this flag allows `pulumi up`
to continue if any errors are encountered.

Currently when we encounter an error while creating/updating a
resource, we cancel the context of the deployment executor, and thus
the deployment stops once the resources that are being processed in
parallel with the failed one finish being updated.

For --continue-on-error, we ignore these errors, and let the
deployment executor continue.  In order for the deployment executor to
exit eventually we also have to mark these steps as done, as the
deployment executor will otherwise just hang, and callers with open
channels waiting for it to finish/report back will hang indefinitely.

The errors in the step will still be reported back to the user by the
OnResourceStepPost callback.

Fixes https://github.com/pulumi/pulumi/issues/14515

### CODE REVIEW

File: pkg/engine/lifecycletest/pulumi_test.go
Comment (Frassle) [inline pkg/engine/lifecycletest/pulumi_test.go:5573]: assert.NoError? Surely this _does_ error now?

File: pkg/engine/lifecycletest/pulumi_test.go
Comment (Frassle) [inline pkg/engine/lifecycletest/pulumi_test.go:5645]: Again shouldn't this error for the update?

File: pkg/resource/deploy/step.go
Comment (Frassle) [inline pkg/resource/deploy/step.go:583]: This can't be safe. We haven't actually managed to update to s.new, we should mark this as done for that state.

File: pkg/engine/lifecycletest/pulumi_test.go
Comment (tgummerer) [reply pkg/engine/lifecycletest/pulumi_test.go:5573]: It doesn't error during the `RegisterResource` call.  The errors are captured in `OnResourceStepPost`, and reported back to the user that way.  An earlier iteration of this that I experimented with did error on `RegisterResource`, but that means the error ends up getting reported to the user twice.  If I understand correctly, `RegisterResource` only errors when there is an error the engine captures, e.g. invalid URNs, etc.

Note that not erroring here when the provider returns an error is also the current behaviour without any `ContinueOnError` flag.

File: pkg/resource/deploy/step.go
Comment (tgummerer) [reply pkg/resource/deploy/step.go:583]: Yeah I'm feeling a bit iffy about this.  It does work, but I'm not sure it's the best way to go, or that it's right.

We did create a `s.new` here, as that's what we initialize the step with, but we didn't fill out any information about the outputs, nor the modified timestamp.  (similarly for `CreateStep` we're not filling in the new ID and not the outputs or timestamps either). Now what done does is unblock the `done` channel [here](https://github.com/pulumi/pulumi/blob/3043adeeb388f31198e0ee42fc708d78c7c0c810/pkg/resource/deploy/source_eval.go#L1864), and we end up returning a valid result to the language SDK.  Meanwhile we are registering the error as mentioned above.

For the current behaviour without `ContinueOnError`, we never close this channel, but we cancel the context of the StepExecutor, so the language SDK just never gets a response back (?? I need to double check that that's what's happing, but I had that impression).  

We could return an error response from `source_eval`, which would also make the `RegisterResource` call above error out, but that makes it throw that error back to the user as well, and we end up with a duplicate error, which I don't think that's the result we want either.  I'll experiment a bit more and will have a look if I can come up with something better.  But any ideas are also welcome.

File: pkg/engine/lifecycletest/pulumi_test.go
Comment (Frassle) [reply pkg/engine/lifecycletest/pulumi_test.go:5573]: > Note that not erroring here when the provider returns an error is also the current behaviour without any

Really... hmm that's worrying. The SDK then thinks this resource is OK and tries to access the outputs but the outputs aren't going to exist?

File: pkg/resource/deploy/step.go
Comment (Frassle) [reply pkg/resource/deploy/step.go:583]: Yeh we have a very similar issue with --targets. 
There's cases where the user program defines a new resource to be created, but we run with targets so it doesn't actually get created but we just return an empty state (I think) back for the resource. I think that causes everything to register as unknown even during a full `up`. 
I don't think any of the SDKs really handle that very well, this looks similar.
I think the right thing to do here is to have a way for register resource to return back that the resource didn't error (at least as far as the user program is concerned) but it also didn't create and so no dependent operations can take place. Might be that's just a case of allowing 'unknown' not just during preview, but during up as well.

File: pkg/engine/lifecycletest/pulumi_test.go
Comment (tgummerer) [reply pkg/engine/lifecycletest/pulumi_test.go:5573]: Oh wait, I think I was confusing something here. we're not returning an error here, but we're also not returning any response.  The whole `programF` will just hang in the non ContinueOnError case, and we're exiting because the context is canceled.  But we're never even hitting this assert if `ContinueOnError` is false.

File: pkg/resource/deploy/step.go
Comment (tgummerer) [reply pkg/resource/deploy/step.go:583]: Ah `--targets` is a good call.  For non-targeted resources there we use a `SameStep`, so the apply actually succeeds, and we re-use the outputs from `old`.  Though I dunno if that makes a lot of sense for updates of failing resources, since the outputs there should really be unknown at that point :thinking:  

Though I'm not even sure we should make the outputs that were previously there for that step unknown, or if we should just return an empty PropertyMap for outputs, since we don't even know if the resource would necessarily have the same outputs as before.

Comment (tgummerer) [reply pkg/resource/deploy/step.go]: I pushed up a potential improvement doing this, handling failures specially, and not returning any outputs the the SDK in that case.

Comment (tgummerer) [reply pkg/resource/deploy/step.go]: In terms of resource application, and what we return to the SDK `SkippedCreateStep` looks like it's the same as `SameStep` to me.  The only difference is that we don't write anything into the snapshot.  

But a failed resource seems kind of different.  I'm not sure we can trust the outputs for those, since we didn't get any back.  For non-targeted resources it seems kind of clear that the outputs haven't changed, but for failed resources this seems like undefined behaviour.  We could parrot back the old outputs, but I'm unsure whether that's better than not returning any outputs, and with that making it clear that the resource is failed and the outputs are not to be trusted?

Comment (Frassle) [reply pkg/resource/deploy/step.go]: > I'm not sure we can trust the outputs for those, since we didn't get any back.

We didn't get any outputs back for a skipped create either, thus why I think it's an iffy part of the system. Let's not replicate that mistake here.

File: pkg/resource/deploy/source_eval.go
Comment (Frassle) [inline pkg/resource/deploy/source_eval.go:1878]: So the SDK gets a response that the resource did register, but just has no outputs, and might have an empty ID?
That seems suspect, I don't trust SDKs are going to handle this well vis-a-vis working out unknown vs undefined vs error'd output states.

File: proto/pulumi/resource.proto
Comment (Frassle) [inline proto/pulumi/resource.proto:165]:
```suggestion
    SkipReason skipReason = 7;                                  // the reason, if any, why the resource creation was skipped.
```

File: sdk/nodejs/output.ts
Comment (Frassle) [inline sdk/nodejs/output.ts:310]: I think to minimise blast radius we could do a pass removing the `ifDryRun` apply logic from nodejs as it's own PR. Nothing _should_ really depend on that, and we can sanity check just that change is safe.

File: cmd/pulumi-test-language/testdata/l2-failed-create-continue-on-error/main.pp
Comment (Frassle) [inline cmd/pulumi-test-language/testdata/l2-failed-create-continue-on-error/main.pp:32]: Feel like this is probably overkill? If `independent` works how would these not work?

Comment (tgummerer) [reply cmd/pulumi-test-language/testdata/l2-failed-create-continue-on-error/main.pp:32]: It's the other way round.  At least in my testing `up` without `--continue-on-error` would sometimes be able to create `independent` and even `independent2`, even when `failing` failed earlier.  So I added these mostly for the purpose of showing that this works now, but doesn't work without `--continue-on-error`.  

We do have engine tests for that as well, so this might be slightly overkill here, but a couple more resources shouldn't make the test too much slower I think, so it felt better to just keep them.

Comment (Frassle) [reply cmd/pulumi-test-language/testdata/l2-failed-create-continue-on-error/main.pp:32]: That's testing race semantics I think, resource registration from the SDKs don't have a defined order unless explicit dependencies are present.

File: pkg/resource/deploy/deployment_executor.go
Comment (tgummerer) [reply pkg/resource/deploy/deployment_executor.go]: It's the exit signal for the workers: https://github.com/pulumi/pulumi/blob/f6b405c24ebc03f293173763ac93a75eacee5d8e/pkg/resource/deploy/step_executor.go#L543

We could maybe do better and have a separate channel to exit the worker, or maybe just re-use the context, but I didn't want to introduce too many additional changes in this PR, as it's already getting somewhat complex.

File: proto/pulumi/resource.proto
Comment (tgummerer) [reply proto/pulumi/resource.proto:165]: Yeah I was struggling with naming this tbh.  Maybe just "Result", and then have the states be success/failed/skipped?  Definitely open to other suggestions.

File: pkg/resource/deploy/source.go
Comment (tgummerer) [reply pkg/resource/deploy/source.go]: Yeah an enum would be much better.  Done in https://github.com/pulumi/pulumi/pull/15740/commits/9ac2fde40639e41703149d618e0037282a3b0d8d

File: proto/pulumi/resource.proto
Comment (tgummerer) [reply proto/pulumi/resource.proto]: Changed this to Result in https://github.com/pulumi/pulumi/pull/15740/commits/3badab9922ce8b0f0e0e0f2cef403baf061c0bbd.

File: pkg/engine/lifecycletest/pulumi_test.go
Comment (Frassle) [inline pkg/engine/lifecycletest/pulumi_test.go:5572]: Should we update RegisterResource here to return the Reason field? Maybe join all the fields into a struct at this point rather than adding another return value, could struct-ify in another PR to keep changes small then just rebase and add Reason to the struct in this PR. I don't like that there's a critical response being returned here that we're not even looking at.

File: pkg/engine/lifecycletest/pulumi_test.go
Comment (Frassle) [inline pkg/engine/lifecycletest/pulumi_test.go:5579]: Should we have another test that does the same as this one but without SupportsResultReporting and check the engine still handles that correctly (for old sdks)

File: pkg/engine/lifecycletest/pulumi_test.go
Comment (Frassle) [inline pkg/engine/lifecycletest/pulumi_test.go:5655]: Same as above, we should have a version of this without SupportsResultReporting for testing old SDKs

File: pkg/resource/deploy/deployment_executor.go
Comment (Frassle) [inline pkg/resource/deploy/deployment_executor.go:452]: Why is the deployment_executor responsible for this rather than the step generator, or even the source eval? We know the step generator/source eval will not see the new registration till the step_executor has finished it because it's dependent.

File: pkg/engine/lifecycletest/pulumi_test.go
Comment (tgummerer) [reply pkg/engine/lifecycletest/pulumi_test.go:5572]: Yeah that sounds like a good idea: done in https://github.com/pulumi/pulumi/pull/15988.

File: pkg/engine/lifecycletest/pulumi_test.go
Comment (tgummerer) [reply pkg/engine/lifecycletest/pulumi_test.go]: Good idea, done for this and below :heavy_check_mark:

File: pkg/engine/lifecycletest/pulumi_test.go
Comment (tgummerer) [reply pkg/engine/lifecycletest/pulumi_test.go]: Done now. I've rebased this PR and temporarily based this PR on that branch to make reviewing easier until that can go through the merge queue.

File: pkg/engine/lifecycletest/pulumi_test.go
Comment (tgummerer) [inline pkg/engine/lifecycletest/pulumi_test.go:5501]: This and the other continue on error tests now live in `continue_on_error_test.go`, as `pulumi_test.go` feels a little bit too much like kitchen sink of tests.

File: pkg/engine/lifecycletest/pulumi_test.go
Comment (tgummerer) [reply pkg/engine/lifecycletest/pulumi_test.go]: The PR merged now, so this is again based on `master`.

**Review by Frassle — APPROVED** [review]

**Pull request merged**
